hips/hip-9999.md

```diff
@@ -0,0 +1,317 @@
+
+Chart-required plugins may need to wait for Chart API v3. hile Helm 3 currently **does** allow additional keys to be added ot the Chart struct while retaining Go backwards-compatibility, any keys that add required functionality for a chart to install would not work with older versions of Helm (such as render plugins) and therefore a top level `plugins` key should wait for Chart API v3.
+
+For discussion:
+Chart API v3 spec for chart-required plugins. There are several plugin types that make sense to be chart-required.
+
+`render` plugin type: While most charts would likely have one render engine, for certain use cases a chart could support multiple filetypes. For example, a chart could have a majority of gottemplate files with one YAMLScript file and others. These could be identified by file type.
+
+`downloader` plugin type: if dependencies are at locations requireing plugins that aren't helm default plugins, then they'll need to be specified too.
+
+```yaml
+plugins:
+    renderer:
+        # made up URLs just for example
+        - render.plugins.helm.sh/pkl@v1.0.0: [".pkl", ".pickle"]
+        - render.plugins.helm.sh/ys@v2.1.1: [".ys"]
+    downloader:
+        - https://artifacthub.io/packages/helm-plugin/s3/s3@0.16.3: ["s3"]
+        - https://artifacthub.io/packages/helm-plugin/git/helm-git: ["git", "git+https", "git+ssh", "git+file"]
+```
+
+### Work in progress
+
+Recent discussions have highlighted the potential of using WebAssembly (WASM) for plugin interfacing. WASM provides a portable, secure, and efficient way to run plugin code across environments, aligning well with Helm’s modular goals. This approach was documented as part of a collaborative effort during KubeCon. Prototyping a proof-of-concept is underway, with a goal of informing the final plugin architecture.
+
+## Backwards compatibility
+
+<!-- Describe potential impact and severity on pre-existing code. -->
+
+To ease the transition from Helm 3 to Helm 4:
+- Existing Helm functionality will continue to work with minimal changes.
+- New plugin-based extensions will be optional but encouraged for advanced use cases.
+- Helm 4 will include migration guides and tools to assist users.
+
+While the Helm commands and SDK may differ, Chart API v2 charts MUST continue to work as-is (see [Requirements](#requirements) above). Therefore, with this new plugin-first architecture, default plugins (required for Chart API v2 charts to work as they did with Helm 3) must be available to Helm 4 users without needing to do additional work.
+
+For discussion:
+- The standard distribution of Helm come bundled with the default plugins for chart api v2 and v3
+- advanced users may optionally install Helm without the default plugins if they know they don't need them, for a more streamlined experience
+- For SDK users:
+    - For SDK users who do not want to allow filesystem access, potentially Go's embed file functioinality could be used for the default Wasm plugins
```

Comment:
Thanks @banjoh for the note about in-memory `embed.FS` and build tags conversation.

```suggestion
    - For SDK users who do not want to allow filesystem access, potentially Go's embed file functionality could be used for the default Wasm plugins. This is a valuable option to explore, especially as `embed.FS` would allow the code to tread either embedded plugins or plugins on disk in the same way. See https://pkg.go.dev/embed. This could also pair nicely with build tags for exclude/include a default plugins directory (eg., default vs contrib directory, etc).
```